ansible_runner/streaming.py

```diff
@@ -0,0 +1,214 @@
+            return obj.hex
+        return json.JSONEncoder.default(self, obj)
+
+
+# List of kwargs options to the run method that should be sent to the remote executor.
+remote_run_options = (
+    'forks',
+    'host_pattern',
+    'ident',
+    'ignore_logging',
+    'inventory',
+    'limit',
+    'module',
+    'module_args',
+    'omit_event_data',
+    'only_failed_event_data',
+    'playbook',
+    'verbosity',
+)
+
+
+class StreamController(object):
+    def __init__(self, control_in, control_out, status_handler=None, event_handler=None,
+                 artifacts_handler=None, cancel_callback=None, finished_callback=None, **kwargs):
+        self.control_in = control_in
+        self.control_out = control_out
+
+        self.kwargs = kwargs
+        self.config = ansible_runner.RunnerConfig(**kwargs)
+        self.status_handler = status_handler
+        self.event_handler = event_handler
+        self.artifacts_handler = artifacts_handler
+
+        self.cancel_callback = cancel_callback
+        self.finished_callback = finished_callback
+
+        self.status = "unstarted"
+        self.rc = None
+
+    def run(self):
+        self.send_job()
+
+        job_events_path = os.path.join(self.config.artifact_dir, 'job_events')
+        if not os.path.exists(job_events_path):
+            os.mkdir(job_events_path, 0o700)
+
+        for line in self.control_in:
+            data = json.loads(line)
+            if 'status' in data:
+                self.status_callback(data)
+            elif 'artifacts' in data:
+                self.artifacts_callback(data)
+            elif 'eof' in data:
+                break
+            else:
+                self.event_callback(data)
+
+        if self.finished_callback is not None:
+            self.finished_callback(self)
+        return self.status, self.rc
+
+    def send_job(self):
+        self.config.prepare()
+        remote_options = {key: value for key, value in self.kwargs.items() if key in remote_run_options}
+
+        buf = io.BytesIO()
+        with zipfile.ZipFile(buf, 'w', compression=zipfile.ZIP_DEFLATED, allowZip64=True) as archive:
+            private_data_dir = self.kwargs.get('private_data_dir', None)
+            if private_data_dir:
+                for dirpath, dirs, files in os.walk(private_data_dir):
+                    relpath = os.path.relpath(dirpath, private_data_dir)
+                    if relpath == ".":
+                        relpath = ""
+                    for fname in files:
+                        archive.write(os.path.join(dirpath, fname), arcname=os.path.join(relpath, fname))
+
+            kwargs = json.dumps(remote_options, cls=UUIDEncoder)
+            archive.writestr('kwargs', kwargs)
+            archive.close()
+        buf.flush()
+
+        data = {
+            'private_data_dir': True,
+            'payload': base64.b64encode(buf.getvalue()).decode('ascii'),
+        }
+        self.control_out.write(json.dumps(data).encode('utf-8'))
```

Comment:
Do we really need to encode the data zip as JSON?  This will considerably increase the size of the data being sent over the wire.

Reply:
This has now been fixed.